sotodlib/preprocess/core.py

```diff
@@ -146,6 +148,126 @@ def register(process_class):
                 f"Preprocess Module of name {name} is already Registered"
             )
 
+def _zeros_cls( item ):
+    """return a callable zeros class that exactly matches type item for use 
+    with wrap_new"""
+    if isinstance( item, np.ndarray):
+        return lambda shape: np.zeros( shape, dtype = item.dtype)
+    elif isinstance( item, RangesMatrix):
+        return RangesMatrix.zeros
+    elif isinstance( item, Ranges ):
+        def temp(shape):
+            assert len(shape) == 1
+            return Ranges( shape[0] )
+        return temp
+    elif isinstance( item, csr_array):
+        return lambda shape: csr_array(tuple(shape), dtype=item.dtype)
+    else:
+        raise ValueError(f"Cannot find zero type for {type(item)}")
+
+def _ranges_matrix_match( o, n, oidx, nidx):
+    """align RangesMatrix n entries to RangesMatrix o"""
+    assert len(oidx)==len(nidx)
+    if len(oidx) > 2:
+        raise NotImplemented
+    for i, x in zip( oidx[0], nidx[0]):
+        o.ranges[i] = _ranges_match( 
+            o.ranges[i], n.ranges[x],
+            [oidx[1]], [nidx[1]]
+        )
+    return o.copy()
+
+def _ranges_match( o, n, oidx, nidx):
+    """align Ranges n to Ranges o"""
+    assert len(oidx)==len(nidx)
+    assert len(oidx)==1
+    omsk = o.mask()
+    nmsk = n.mask()
+    omsk[oidx[0]] = nmsk[nidx[0]]
+    return Ranges.from_mask(omsk)
+
+def _expand(new, full, wrap_valid=True):
+    """new will become a top level axismanager in full once it is matched to
+    size"""
+    if 'dets' in new._axes:
+        _, fs_dets, ns_dets = full.dets.intersection(
+            new.dets, 
+            return_slices=True
+        )
+    if 'samps' in new._axes:
+        _, fs_samps, ns_samps = full.samps.intersection(
+            new.samps, 
+            return_slices=True
+        )
+    else:
+        fs_samps = slice(None)
+
+    out = core.AxisManager()
+    for k, v in full._axes.items():
+        if k in list(new._axes.keys())+['dets','samps']:
+            out._axes[k] = v 
+
+    for a in new._axes:
+        if a not in out:
+            out.add_axis( new[a] )
```

Comment:
wouldn't this all just be this?:
````
for a in new._axes:
    if a in full._axes:
        out.add_axis(full[a])
    else:
        out.add_axis(new[a])
````

Reply:
Hey @kmharrington, yes I tried this at first then ran into an error because the `valid` flag needs to have `dets` and `samps` but some of the processes don't have both of those axes so I thought it was easier to just force `out` to always have `dets` and `samps`. Below is the traceback when I run with your suggestion for reference: 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File <timed exec>:1

File ~/.local/lib/python3.10/site-packages/sotodlib/preprocess/core.py:379, in Pipeline.run(self, aman, proc_aman, select)
    377 if run_calc:
    378     process.calc_and_save(aman, proc_aman)
--> 379     update_full_aman( proc_aman, full, self.wrap_valid)
    381 if select:
    382     process.select(aman, proc_aman)

File ~/.local/lib/python3.10/site-packages/sotodlib/preprocess/core.py:268, in update_full_aman(proc_aman, full, wrap_valid)
    264 if fld not in full._fields:
    265     assert isinstance(proc_aman[fld], core.AxisManager)
    266     full.wrap( 
    267         fld,
--> 268         _expand( proc_aman[fld], full, wrap_valid=wrap_valid)
    269     )

File ~/.local/lib/python3.10/site-packages/sotodlib/preprocess/core.py:247, in _expand(new, full, wrap_valid)
    242     v = Ranges.from_mask(m)
    244     valid = RangesMatrix( 
    245         [v if i in fs_dets else x for i in range(full.dets.count)]
    246     )
--> 247     out.wrap('valid',valid,[(0,'dets'),(1,'samps')])
    248 return out

File ~/.local/lib/python3.10/site-packages/sotodlib/core/axisman.py:605, in AxisManager.wrap(self, name, data, axis_map, accept_new, accept_merge)
    602 if not isinstance(axis, AxisInterface):
    603     # So it better be a string label... that we've heard of.
    604     if axis not in self._axes:
--> 605         raise ValueError("Axis assignment refers to unknown "
    606                          "axis '%s'." % axis)
    607     axis = self._axes[axis]
    608 axis = axis.resolve(data, index)

ValueError: Axis assignment refers to unknown axis 'samps'.
```

Reply:
ahh gotcha. makes sense!